src/library/scala/collection/Seq.scala

```diff
@@ -793,19 +802,17 @@ trait SeqOps[+A, +CC[_], +C] extends Any
     *                If an element value `x` appears
     *                ''n'' times in `that`, then the first ''n'' occurrences of `x` will not form
     *                part of the result, but any following occurrences will.
-    *  $willNotTerminateInf
-    *  $willForceEvaluation
     */
   def diff(that: Seq[_ >: A]): C = {
     val occ = occCounts(that)
-    //TODO make diff and intersect use efficient lazy implementations now that fromSpecific accepts an IterableOnce
-    val b = newSpecificBuilder
-    for (x <- this) {
+    fromSpecific(iterator.filter { x =>
```

Comment:
does this need to use `iterator`, or can it directly use `filter`?

Reply:
We should also override these methods in `StrictOptimizedSeqOps` with the builder-based implementations, for performance.

Reply:
if it doesn't use `fromSpecific(iterator.filter {...})`, then if `diff` is called on a `View`, it will re-evaluate `filter` every time, but it will not reconstruct `occ`. I'm not sure if there is a better solution to this.

Reply:
It's also tough to override in `StrictOptimizedSeqOps`, as `occCounts` is `private[this]`. Should it be changed to be `protected[collection]`?

Reply:
> if diff is called on a View, it will re-evaluate filter every time, but it will not reconstruct occ. I'm not sure if there is a better solution to this.

I don’t understand, can you elaborate, please?

> Should it be changed to be protected[collection]?

Yes, I think so.

Reply:
Ideally, we would just call `filter`, and could leave out the indirection of using an iterator and calling `fromSpecific`. However, calling `View#filter` produces a `View` whose `iterator` method calls `Iterator#filter`. Each call to `iterator` on the filtered `View` will create a new iterator; however, consuming the first iterator will (presumably) remove values from the map of counts used by the predicate. That map is never repopulated, so subsequent calls to `View#iterator` will yield iterators containing the wrong result.

Reply:
I see. Indeed, the fact that we do a side-effect within the predicate of `filter` makes things harder to get right. That being said, I think the problem is also present in the version you propose: if `fromSpecific` is lazy and doesn’t memoize its elements, then the `iterator.filter` expression will be evaluated more than once, leading to wrong results, no?

Reply:
@julienrf I have added overrides in `StrictOptimizedSeqOps`